squirrel/plugin.py

```diff
@@ -18,30 +18,65 @@ def load_module():
         return importlib.import_module(f'squirrel.plugins.{project_type}')
 
     @staticmethod
-    def get_files(path):
-        find_output = subprocess.run(
-            f'find {path} -type f -not -path "*/[@.]*"',
-            shell=True,
-            capture_output=True,
-            text=True)
-        return find_output.stdout.strip().split('\n')
+    def get_files(path, ignores):
+        """Function to find all non-ignored files i project directory"""
+        ignores_ext = tuple(i[1:] for i in ignores.get('ext'))
+        ignores_file = tuple(ignores.get('file'))
+        ignores_dir = tuple(i[:-1] for i in ignores.get('dir_full'))
+        project_files = []
+        for file in glob('**/*', recursive=True):
+            file_path = os.path.join(path, file)
+            if not os.path.isdir(file_path):
+                if not file_path.endswith(ignores_ext) \
+                        and file_path.rsplit('/', 1)[1] not in ignores_file \
+                        and file_path.rsplit('/', 1)[0] not in ignores_dir:
+                    project_files.append(file_path)
+        return project_files
+
+    @staticmethod
+    def import_ignores(wd, file, logger):
+        """Function to read ignore file and store extensions, dir and files
+         to a dictionary"""
+        ignores = {
+            'ext': [],
+            'dir_full': [],
+            'file': []
+            }
```

Comment:
With `glob`, I don't think it's necessary to distinguish between ext from file. we just need to pass the "expression" to glob and it will return the right files. In the case of a full directory (e.g `.some_dir/`, etc) we could add an `*` and pass it to `glob`.

Reply:
Haven't got it to work when we try to exclude files/extentions. I'll try to minimize the `if not in` statements.
Do you have an example on how to solve it?

Reply:
I was thinking you could use `glob` inside the ignore function like so:
```python
>>> from glob import glob
>>> ignores = {
...  'dir': [],
...  'files': ['.squirrel/*', '*.ini', 'test.txt'],
... }
>>> to_ignore = []
>>> for file in ignores['files']:
...     to_ignore.append(glob(file))
>>> to_ignore
[['.squirrel/ignore', '.squirrel/squirrel-project.xml', '.squirrel/watch-data.xml'], ['tox.ini'], []]
```

Reply:
Just so I understand. You mean use glob inside `import_ignores() `and store the files to a set and then again in `get_files()`. Then use `set.difference()` to extract the files we want to count?

Reply:
Yeah get all the file paths using glob from the patters (`test.txt`, `*.ini`, etc) in the ignore fille.

>store the files to a set and then again in get_files(). Then use set.difference() to extract the files we want to count?

As you wish, you can keep it as a list.

here's and example of how the function could be:
```Python
    @staticmethod
 34     def import_ignores(wd, file, logger):
 35         """Function to read ignore file and store extensions, dir and files
 36          to a dictionary"""
 37         ignores = {
 38             'dir': [],
 39             'file': []
 40             }
 41         try:
 42             with open(file, 'r') as file:
 43                 for line in file.readlines():
 44                     add_line = line.strip()
 45                     if add_line.startswith('#') or add_line == '':
 46                         continue
 47                     elif add_line.endswith('/'):
 48                         ignores['dir'].append(''.join(f'{wd}/{add_line}'))
 49                     else:
 50                         ignores['file'].extend(glob(add_line))
 51         except FileNotFoundError:
 52             logger.debug(f'{__name__} No ignore file found <{file}>')
 53         return ignores
```

Reply:
Here's my `import_ignores()` atm:

We need `dir` and `file` these are used by `Handler()` to sort out ignored files when `watch` is looking for modified files.
`glob()` make a list of lists. You'll see my comprihension to fix it at the end of the function

```python
    def import_ignores(wd, file, logger):
        """Function to read ignore file and store extensions, dir and files
         to a dictionary"""
        ignores = {
            'dir': [],
            'file': [],
            'ignore': set(),
            }
        try:
            ignore = []
            ignore_set = set()
            with open(file, 'r') as file:
                for line in file.readlines():
                    add_line = line.strip()
                    if add_line.startswith('#') or add_line == '':
                        continue
                    elif add_line.endswith('/'):
                        ignores['dir'].append(''.join(f'{wd}/{add_line}'))
                        ignore.append(
                            glob(f'{add_line}*', recursive=True))
                    else:
                        ignores['file'].append(add_line)
                        ignore.append(
                            glob(f'**/{add_line}', recursive=True))
            # Comprihension to store list of lists into a single set
            {ignore_set.update(f) for f in ignore}
            # Comprihension to add full path to file for use in get_files()
            {ignores['ignore'].add(f'{wd}/{f}') for f in ignore_set}
        except FileNotFoundError:
            logger.debug(f'{__name__} No ignore file found <{file}>')
        return ignores
```

Reply:
Thanks, this implementation is great. There's just somethings that would benefit changing.

```Python
elif add_line.endswith('/'):
      ignores['dir'].append(''.join(f'{wd}/{add_line}'))
      ignore.append(
                  glob(f'{add_line}*', recursive=True))
```
I don't think it should be recursive here as most users will probably not expect such a behaviour. The way you're using the glob expression here would only target the all the files inside the dir and not the dir itself so we should probably keep the dir on its own.

```Python
else:
      ignores['file'].append(add_line)
      ignore.append(
                   glob(f'**/{add_line}', recursive=True))
```
If you want to avoid having nested lists in ignore, you can use `ignore.extend()` instead of append. Here I would also say that we probably should drop `**/` and `recursive=True` as the behaviour would be unexpected.

Reply:
> I don't think it should be recursive here as most users will probably not expect such a behaviour. The way you're using the glob expression here would only target the all the files inside the dir and not the dir itself so we should probably keep the dir on its own.

The tought behind this is to add all files in the ignored dir into the set so we don't need  a `if ignored_dir` expresion in `get_files()` but let `difference_update()` remove all ignores for project_files.

Shall i still change it? More pythonic an readable to have a `if ignored_dir` expresion?